### PR TITLE
[JENKINS-35881] replay a run

### PR DESCRIPTION
**Pull request by imeredith**
This PR allows a run to be replayed

### CODE REVIEW

**Comment by michaelneale** [comment]
pretty much all the multibranch tests fail with this change

**Comment by imeredith** [comment]
@michaelneale should be fixed now

**Comment by imeredith** [comment]
Still need to document, doing that now

File: blueocean-rest/README.md [hunk 1]
```diff
@@ -337,6 +337,22 @@ Pipelines can be nested inside folder.
       "qeueudTime" : "2016-06-22T11:05:41.309+1200"
     }
 
+## Replay a pipeline run
```
Comment (michaelneale) [inline blueocean-rest/README.md]: Can it mention here if it runs the exact commit or HEAD? (I have already forgotten)

File: blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/AbstractRunImplTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    @Rule
+    public GitSampleRepoRule sampleRepo = new GitSampleRepoRule();
+
+    @Before
+    public void setup() throws Exception{
+        super.setup();
+        sampleRepo.init();
+    }
+    //Disabled, see JENKINS-36453
+    //@Test
```
Comment (michaelneale) [inline blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/AbstractRunImplTest.java]: can we have the test use a multibranch (as we know that works?)

**Comment by michaelneale** [comment]
Looks good, but would like to see the test enabled with multibranch if possible (as we know that works).

**Comment by michaelneale** [comment]
I got a "no crumbs included" 🐛  when trying: 

```
curl -XPOST http://localhost:8080/jenkins/blue/rest/organizations/jenkins/pipelines/bee/branches/master/runs/1/replay
```

**Comment by blueoceandeploy** [comment]
Ah it needs to be Content-Type: application json

**Comment by michaelneale** [comment]
that test run failed (odd error, looks like process spawning).

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BlueRun.java [hunk 3]
```diff
@@ -193,8 +193,14 @@ public final String getEndTimeString(){
     @Navigable
     public abstract Object getLog();
 
+    /**
+     * Replays a pipeline. Sam SCM commit/revision should be used. in the new run.
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BlueRun.java]: Typo? s/Sam/Same/

File: blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/AbstractRunImplTest.java [hunk 4]
```diff
@@ -60,8 +68,53 @@ public void replayRunTest() throws Exception {
...
+
+        Assert.assertNotEquals(new PipelineRunImpl(b1, null).getCommitId(), new PipelineRunImpl(b2, null).getCommitId());
+
+        request().post("/organizations/jenkins/pipelines/p/branches/master/runs/1/replay").build(String.class);
+
+        j.waitForCompletion(job1.getLastBuild());
 
+        request().get("/organizations/jenkins/pipelines/p/branches/master/runs/3/").build(Map.class);
+        request().get("/organizations/jenkins/pipelines/p/branches/master/runs/3/").build(Map.class);
+        Map r = request().get("/organizations/jenkins/pipelines/p/branches/master/runs/3/").build(Map.class);
```
Comment (vivek) [inline blueocean-rest-impl/src/test/java/io/jenkins/blueocean/service/embedded/AbstractRunImplTest.java]: There are similar looking /runs/3 GET calls, is that intended?

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BlueRun.java [hunk 5]
```diff
@@ -193,8 +193,14 @@ public final String getEndTimeString(){
...
+     *
+     * @return The queued item.
+     */
     @POST @TreeResponse @WebMethod(name = "replay")
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BlueRun.java]: From REST API perspective: I see stop, replay verbs appearing in the API URL. Could we not use POST .../runs/:id with payload to indicate action (stop, replay)? Also it fits more of PUT than POST. PUT is considered idempotent, stop and replay both are in fact idempotent. 

```
POST /.../runs/1
{
  "action":"replay"
}
```

**Comment by vivek** [comment]
@imeredith Looks good in general, left some comments.

**Comment by vivek** [comment]
LGTM 🐝 As per discussion with @imeredith, we are going to defer addressing verb in URL for later time when there is yet another verb for run needed.

**Pull request merged**